### PR TITLE
Implement statistics to integrate into query planner

### PR DESCRIPTION
**Pull request by lriuui0x0**
## What is the goal of this PR?

We want to Implement statistics for query planner to generate optimised query plan, including the count for various of types of schema objects:
    - vertex count for a particular type
    - transitive (up to subtypes) count for a particular type
    - `has` edge count between a thing type and an attribute type
    - total `has` edge count between a root type and an attribute type

The overall strategy is divided into two categories:
    - for entity/relation/role vertex count, we hold the delta of count change, which will be merged using rocksdb's `mergeUntracked` at commit time.
    - for attribute vertex/has edge count, we commit an update work item which will be picked up by a statistics updater thread to update the statistics count. The updater thread will inspect the work item, and check if we have counted this item before and modify the count accordingly.
       The reason we use a different strategy for attribute vertex/has edge count is because they use a `put` call to rocksdb, which makes it impossible to know if the `put` will result in `+1` for count or `+0` for count.

## What are the changes implemented in this PR?

- Added `STATISTICS_THINGS` prefix to hold the instance count statistics.
     - For vertex count the format will be `STATISTICS_THINGS <type_iid> VERTEX_COUNT`
     - For transitive vertex count the format will be `STATISTICS_THINGS <type_iid> VERTEX_TRANSITIVE_COUNT`
     - For has edge count the format will be `STATISTICS_THINGS <from_type_iid> HAS_EDGE_COUNT <to_type_iid>`
     - For total has edge count the format will be `STATISTICS_THINGS <from_type_iid> HAS_EDGE_TOTAL_COUNT <to_type_iid>`
- Added `STATISTICS_COUNT_JOB` prefix to hold the work item that the statistics updater thread will inspect and correct count.
    - For attribute vertex count the format will be `STATISTICS_COUNT_JOB ATTRIBUTE_VERTEX <att_iid>`
    - For has edge count the format will be `STATISTICS_COUNT_JOB HAS_EDGE <thing_iid> <att_iid>`
- Added `STATISTICS_COUNTED` prefix to hold the if a statistics item has been counted.
    - For attribute vertex count the format will be `STATISTICS_COUNTED ATTRIBUTE_VERTEX <att_iid>`
    - For has edge count the format will be `STATISTICS_COUNTED HAS_EDGE <thing_iid> <att_iid>`
- Added a `Statistics` inner class to the `DataGraph` which will hold thing statistics for the current transaction.
- Added the following statistics modification call for the graph to notify statistics that graph entities have been changed:
    - `vertexCreated/vertexDeleted(typeIId)` for a non-attribute vertex has been created/deleted
    - `attributeVertexCreated/attributeVertexDeleted(attIID)` for an attribute vertex has been (potentially) created/deleted.
    - `hasEdgeCreated/hasEdgeDeleted(thingIID, attIID)` for a has edge has been (potentially) created/deleted.
- Commit statistics when data graph is committed.
- When there're attributes and has edge updates, we notify the statistics updater thread (`StatisticsUpdater`) to inspect all update works and correct the statistics.

### CODE REVIEW

**Review by lolski — APPROVED** [review]

File: rocks/RocksDatabase.java [hunk 1]
```diff
@@ -267,4 +279,65 @@ public SchemaGraph schemaGraph() {
...
+
+    static class StatisticsBackgroundCounter {
+        private final RocksSession.Data session;
+        private final Thread thread;
+        private final Semaphore countJobNotifications;
+        private boolean isStopped;
+
+        StatisticsBackgroundCounter(RocksSession.Data session) {
+            this.session = session;
+            // NOTE: We use a Semaphore instead of CountDownLatch here because semaphore is closer to our usage pattern here.
```
Comment (lolski) [inline rocks/RocksDatabase.java:291]: Let's remove this comment

File: graph/SchemaGraph.java [hunk 2]
```diff
@@ -414,16 +418,19 @@ public boolean isModified() {
         return isModified;
     }
 
+    // TODO: verify
```
Comment (lolski) [inline graph/SchemaGraph.java:421]: Let's remove this line

File: rocks/RocksDatabase.java (hunk 1 above)
Comment (lriuui0x0) [reply rocks/RocksDatabase.java:291]: Oh I thought you want to put that in? @lolski

Comment (lolski) [reply rocks/RocksDatabase.java:291]: I meant to add the comment in the PR, not in the code @lriuui0x0

Comment (lolski) [reply rocks/RocksDatabase.java:291]: I'm adding it here:

**Pull request merged**